File: Data Structures/Practice_in_Python/Queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    def enqueue(self, value):
        newNode = Node(value)
        if (self.length == 0):
            self.first = newNode
            self.last = self.first
        else:
            self.last.next = newNode
            self.last = newNode
        self.length += 1
        return self

    def dequeue(self):
        oldFirst = self.first
        newFirst = oldFirst.next
        self.first = newFirst
        oldFirst.next = None
        self.length -= 1
        return oldFirst.value

    def log(self):
        data = []
        current = self.first
        for _ in range(self.length):
            data.append(current.value)
            current = current.next
        print(data)


q = Queue()
q.enqueue(10)
q.enqueue(20)
q.enqueue(30)
q.log()
print(q.dequeue())
print(q.dequeue())
print(q.dequeue())
q.log()

File: Fundamentals/Data Structures/queue.js
class Node {
	constructor(value) {
		this.value = value;
		this.next = null;
	}
}


/**
 
 Queue => FIFO Data Structure
 ============================
 


 Time Complexity
 ===============
 Insertion - O(1)

 Removal - O(1)

 Searching - O(N)

 Access - O(N)

 */
class Queue {
	constructor() {
		this.first = null;
		this.last = null;
		this.length = 0;
	}
	enqueue(value) {
		let newNode = new Node(value);
		if (this.length === 0) {
			this.first = newNode;
			this.last = this.first;
		} else {
			this.last.next = newNode;
			this.last = newNode;
		}
		this.length++;
		return this.length;
	}
	dequeue() {
		if (this.length === 0) return null;
		let removed = this.first;
		if (this.length === 1) {
			this.first = null;
			this.last = null;
		} else {
			this.first = this.first.next;
		}
		this.length--;
		return removed.value;
	}
	print() {
		let arr = [];
		let temp = this.first;
		for (let i = 0; i < this.length; i++) {
			arr.push(temp.value);
			temp = temp.next;
		}
		console.log(arr);
	}
}

let q = new Queue();
q.enqueue(10);
q.enqueue(20);
q.enqueue(30);
console.log(q.dequeue());
// console.log(q.dequeue());
// console.log(q.dequeue());
// console.log(q.dequeue());
q.print();
console.log(q);